notebooks/_06-Architecture.md

```diff
@@ -0,0 +1,429 @@
+
+### 10.2 How much memory and time will it take the system to execute?
+- **Offline / online prediction**
+- **Model training**
+Any personal laptop should be able to execute the system. We don't envision the need for large memory requirements.
+It is unknown exactly how long it will take the system to run, but the objective is to have each sub-system take as little time as possible.
+
+### 10.3 How will the system interact with the external world to acquire resources? (e.g. get data or spawn virtual machines
+We can access data directly from a specified weblink or path for the dataset within the established repo.
+
+#### 10.3.1 How does the system decide how much of an external resource is needed? (e.g. how many virtual CPUs are required during training?) 
+N/A
+
+## 11. Description of how the system will scale:
+### 11.1 With increasing dataset sizes: 
+There is no current plan for scaling the system to adapt to large datasets.
+
+### 11.2 With increasing numbers of models being used:
+The number of models used should not affect the scalability of the project.
+
+### 11.3 With increasing numbers of parameters:
+N/A
+
+### 11.4 With increasing numbers of divergences/losses:
+N/A
+
+## 12. Description of how the system will interface with external systems:
+- **E.g. if a model needs to be served in Go but it was trained in Python…**
+- **If the system needs to manipulate Kubernetes for training...**
+
+N/A
+
+## 13. Description of how errors will be handled:
+### 13.1 What are common expected errors from users and how can we guard against them?
+Having a wrong specification for a desired model, pylogit has built-in mechanisms to guard against it
+Having a wrong specification for a causal graph, we can guard fro
+
+### 13.2 Will we try to fix errors or merely notify users of the error’s presence?
+We will only notify users of the error's presence
+
+### 13.3 When errors are encountered, will we quit immediately or wait until some specified point before notifying users of errors? For which errors is each strategy appropriate?
+Quit immediately in the case of all errors.
+
+### 13.4 What are the conventions for error messages that the system reports?
+
+
+### 13.5 Where are errors processed? At the point of detection, by a central error handling class, by functions above in the call stack, etc.
+Errors are processed at the point of detection.
+
+### 13.6 What level of responsibility does each class have in validating its own input? Is there a central class (or set of classes) that performs all validation? When can classes assume clean information?
+Code for input validation can be built in separate functions and be used part of analysis functions as seen needed.
+
+### 13.7 How will we use (or not use) type hinting to help prevent errors?
+Docstrings will suggest to the users what each of the function takes in as types and what it returns.
```

Comment:
```suggestion
Docstrings will suggest to the users what each of the functions takes in as types and what it returns.
```

Reply:
Also, I would suggest adding explicit [mypy](http://mypy-lang.org/) typing hints to our docstrings.

I found mypy to be incredibly helpful in past projects for catching type errors that may have gone unnoticed due to lack of sufficiently stringent tests or lack of input validation methods.